File: typescript/problem_53.ts
export class StackQueue {
    private inbox: number[] = [];
    private outbox: number[] = [];
    enqueue(val: number): void { this.inbox.push(val); }
    dequeue(): number | null {
        if (!this.outbox.length) while (this.inbox.length) this.outbox.push(this.inbox.pop()!);
        return this.outbox.length ? this.outbox.pop()! : null;
    }
}

export function problem53(): void {
    console.log("Problem 53:");
    const q = new StackQueue();
    q.enqueue(1); q.enqueue(2); q.enqueue(3);
    if (q.dequeue() !== 1) throw new Error("assertion failed");
    if (q.dequeue() !== 2) throw new Error("assertion failed");
    q.enqueue(4);
    if (q.dequeue() !== 3) throw new Error("assertion failed");
    if (q.dequeue() !== 4) throw new Error("assertion failed");
    if (q.dequeue() !== null) throw new Error("assertion failed");
    console.log("All Tests Passed");
}

File: rust/problem_53.rs
pub struct StackQueue {
    inbox: Vec<i32>,
    outbox: Vec<i32>,
}

impl StackQueue {
    pub fn new() -> Self { StackQueue { inbox: vec![], outbox: vec![] } }
    pub fn enqueue(&mut self, v: i32) { self.inbox.push(v); }
    pub fn dequeue(&mut self) -> Option<i32> {
        if self.outbox.is_empty() {
            while let Some(v) = self.inbox.pop() { self.outbox.push(v); }
        }
        self.outbox.pop()
    }
}

pub fn problem53() {
    println!("Problem 53:");
    let mut q = StackQueue::new();
    q.enqueue(1); q.enqueue(2); q.enqueue(3);
    assert_eq!(q.dequeue(), Some(1));
    assert_eq!(q.dequeue(), Some(2));
    q.enqueue(4);
    assert_eq!(q.dequeue(), Some(3));
    assert_eq!(q.dequeue(), Some(4));
    assert_eq!(q.dequeue(), None);
    println!("All Tests Passed");
}
